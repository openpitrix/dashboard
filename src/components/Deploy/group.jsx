import React from 'react';
import PropTypes from 'prop-types';
import _ from 'lodash';

import factory from 'lib/config-parser/factory';
import Section from './section';

import styles from './index.scss';

export default class DeployGroup extends React.Component {
  static propTypes = {
    detail: PropTypes.object,
    onChange: PropTypes.func,
    onEmpty: PropTypes.func,
    seq: PropTypes.number
  };

  static defaultProps = {
    detail: {},
    seq: 0,
    onChange: _.noop,
    onEmpty: _.noop
  };

  renderSingle = item => {
    if (typeof item.toJSON === 'function') {
      item = item.toJSON();
    }

    return (
      <Section
        detail={item}
        onChange={this.props.onChange}
        onEmpty={this.props.onEmpty}
      />
    );
  };

  render() {
    const { detail, seq } = this.props;
    const isNormalGroup = Array.isArray(detail.items);
    const title = isNormalGroup ? detail.title : detail.label || detail.key;
    const items = isNormalGroup ? detail.items : factory(detail.properties);

    // inject label prefix
    let titlePrefix = '';
    if (detail.labelPrefix) {
      titlePrefix += `${_.capitalize(detail.labelPrefix)}: `;
    }

    return (
      <div className={styles.group}>
        <h2 className={styles.title}>{`${seq + 1}. ${titlePrefix}${title}`}</h2>
        {items.map((item, idx) => (
          <div key={idx}>{this.renderSingle(item)}</div>
        ))}
      </div>
    );
  }
}
