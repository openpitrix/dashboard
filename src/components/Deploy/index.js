import Section from './section';

export Group from './group';
export default Section;
