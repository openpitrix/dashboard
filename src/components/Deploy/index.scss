@import "~scss/vars";

.group {
  margin-bottom: 60px;

  .title {
    display: block;
    margin-bottom: 32px;
    font-weight: normal;
    @include BodyRobotoLeftBlack;
  }
}

.row {
  display: flex;
  margin-bottom: 20px;

  .label{
    display: inline-block;
    float: left;
    margin-right: 24px;
    margin-top: 8px;
    width: 168px;
    line-height: 1.2;
    font-size: 14px;
    font-weight: 500;
    color: $N500;
    text-align: right;
    word-break: break-word;
  }

  .item {
    max-width: 488px;
  }

  .tips{
    font-size: 12px;
    color: $N100;
    display: inline-block;
    margin-top: 6px;
    width: 360px;
    word-break: break-word;
  }
  .radio {

  }
  .input {
    width: 256px;
  }

  .select {

  }
  .slider{
    width: 256px;
    margin-right: 12px;
  }
  .inputSmall{
    max-width: 4em;
    padding: 7px 5px;
  }
  .sliderInput{
    display: inline-block;
  }
  .text {

  }
}

// fix
.isRadio {
  margin-top: 0 !important;
}

.editorWrap {
  width: 90%;
  max-width: 95% !important;
  padding-right: 10px;
}

.isYamlLabel {
  width: auto;
  word-break: normal !important;
  color: $N500;
}