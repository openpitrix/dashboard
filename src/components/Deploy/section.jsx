import React, { Fragment } from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';
import _ from 'lodash';

import {
  Radio, Input, Select, Slider, CodeMirror
} from 'components/Base';

import styles from './index.scss';

export default class Section extends React.Component {
  static propTypes = {
    className: PropTypes.string,
    detail: PropTypes.shape({
      defaultValue: PropTypes.oneOfType([
        PropTypes.string,
        PropTypes.number,
        PropTypes.bool
      ]),
      description: PropTypes.string,
      keyName: PropTypes.string.isRequired, // key is reserved by react
      label: PropTypes.string,
      required: PropTypes.bool,
      type: PropTypes.string.isRequired,
      renderType: PropTypes.string,
      items: PropTypes.array
    }),
    onChange: PropTypes.func,
    onEmpty: PropTypes.func
  };

  static defaultProps = {
    detail: {
      description: '',
      keyName: '',
      label: '',
      required: false,
      type: 'string'
    }
  };

  state = {
    value: null
  };

  handleChange = val => {
    const { keyName } = this.props.detail;

    if (typeof val === 'object' && val.nativeEvent) {
      val = val.currentTarget.value;
    }

    this.setState({ value: val });

    this.props.onChange(keyName, val);
  };

  getDefaultValue() {
    const { defaultValue, ...rest } = this.props.detail;

    if (!defaultValue) {
      if (Array.isArray(rest.range)) {
        return _.isObject(rest.range[0])
          ? _.get(rest.range[0], 'value')
          : rest.range[0];
      }
      if (Array.isArray(rest.items)) {
        return _.get(rest.items[0], 'value');
      }
    }

    return defaultValue;
  }

  getValue() {
    const { value } = this.state;
    return value !== null ? value : this.getDefaultValue();
  }

  getNumbericValue() {
    const val = parseInt(this.getValue());
    return Number.isNaN(val) ? 0 : val;
  }

  formatLabel = (value, key) => {
    let name = '';
    switch (key) {
      case 'cpu':
        name = `${value} Core`;
        break;
      case 'memory':
        name = `${value / 1024} GB`;
        break;
      case 'instance_class':
        name = value === 0 ? 'High Performance' : 'Super-high Performance';
        break;
      default:
        name = value.toString();
        break;
    }
    return name;
  };

  renderLabel() {
    const { keyName, label, renderType } = this.props.detail;

    return (
      <label
        className={classnames(styles.label, {
          [styles.isRadio]: renderType === 'radio',
          [styles.isYamlLabel]: renderType === 'yaml'
        })}
      >
        {label || keyName}
      </label>
    );
  }

  renderItem() {
    const {
      renderType,
      description,
      defaultValue,
      keyName,
      originKey,
      items,
      required,
      ...rest
    } = this.props.detail;

    // hook: render content for empty items
    if (renderType === 'radio' && _.isEmpty(rest.range)) {
      return this.props.onEmpty(keyName);
    }

    let content = null;

    switch (renderType) {
      case 'input':
        content = (
          <Fragment>
            <Input
              className={styles.input}
              name={keyName}
              type="text"
              defaultValue={defaultValue}
              maxLength={50}
              onChange={this.handleChange}
              required={required}
            />
            <p className={styles.tips}>{description}</p>
          </Fragment>
        );
        break;

      case 'number':
        content = (
          <Fragment>
            <Input
              className={styles.input}
              name={keyName}
              type="number"
              defaultValue={defaultValue}
              onChange={this.handleChange}
              min={rest.min}
              max={rest.max}
              required={required}
            />
            <p className={styles.tips}>
              Range: {rest.min} - {rest.max}
            </p>
          </Fragment>
        );
        break;

      case 'radio':
        content = (
          <Radio.Group
            className={styles.radio}
            onChange={this.handleChange}
            name={keyName}
            value={this.getValue()}
          >
            {rest.range.map((item, idx) => {
              let value = '';
              if (_.isObject(item)) {
                value = item.value;
              } else {
                value = item;
              }
              const name = keyName === 'cluster.runtime'
                ? item.name
                : this.formatLabel(value, originKey);

              return (
                <Radio value={value} key={idx}>
                  {name}
                </Radio>
              );
            })}
          </Radio.Group>
        );
        break;

      case 'select':
        content = (
          <Select
            className={styles.select}
            onChange={this.handleChange}
            disabled={items.length === 0}
            name={keyName}
            value={this.getValue()}
          >
            {items.map(({ name, value }, idx) => (
              <Select.Option key={idx} value={value}>
                {`${name}`}
              </Select.Option>
            ))}
          </Select>
        );
        break;

      case 'text':
        content = (
          <textarea
            className={styles.text}
            name={keyName}
            defaultValue={defaultValue}
            maxLength={1000}
            onChange={this.handleChange}
          />
        );
        break;

      case 'slider':
        content = (
          <Fragment>
            <Slider
              min={rest.min}
              max={rest.max}
              step={rest.step}
              value={this.getNumbericValue()}
              onChange={this.handleChange}
              className={styles.slider}
            />
            <span className={styles.sliderInput}>
              <Input
                className={styles.inputSmall}
                type="number"
                name={keyName}
                value={this.getValue()}
                min={rest.min}
                max={rest.max}
                onChange={this.handleChange}
              />{' '}
              GB
            </span>
            <p className={styles.tips}>
              {`${rest.min}GB - ${
                rest.max
              }GB, The volume size for each instance`}
            </p>
          </Fragment>
        );
        break;

      case 'yaml':
        content = rest.yaml && (
          <CodeMirror
            code={rest.yaml}
            onChange={this.handleChange}
            name={keyName}
          />
        );
        break;

      default:
        break;
    }

    return content;
  }

  render() {
    const { className, detail } = this.props;

    return (
      <div className={classnames(styles.row, className)}>
        {this.renderLabel()}
        <div
          className={classnames(styles.item, {
            [styles.editorWrap]: detail.renderType === 'yaml'
          })}
        >
          {this.renderItem()}
        </div>
      </div>
    );
  }
}
