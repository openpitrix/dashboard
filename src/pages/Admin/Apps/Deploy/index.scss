@import "~scss/vars";

.header {
  margin: 24px 0;
  a {
    line-height: 20px;
    font-weight: 500;
    color: $title-color;
    transition: color .3s ease-in-out;
    &:hover {
      color: lighten($title-color, 20%);
    }
  }
}

.createForm {
  display: block;
  margin: 0;
  margin-top: 5px;

  .moduleTitle{
    //margin: 50px 0 25px 0;
    margin-bottom: 28px;
    font-size: 16px;
    color: $N500;
    &:not(:first-of-type) {
      margin-top: 50px;
    }
  }
  .cellModule{
    margin-bottom: 20px;
    clear: both;
  }
  .name{
    display: inline-block;
    float: left;
    margin-right: 24px;
    width: 168px;
    font-size: 14px;
    font-weight: 500;
    color: $N500;
    text-align: right;
  }
  .showWord{
    display: inline-block;
    max-width: 420px;
    line-height: 32px;
  }

  .radio,.checkbox{
    margin-right: 20px;
  }
  .add{
    display: inline-block;
    font-weight: bold;
  }
  .select{
    width: 256px;
    margin-right: 12px;
  }
  .textarea{
    box-sizing: border-box;
    padding: 7px 10px;
    width: 400px;
    height: 88px;
    border-radius: 2px;
    background-color: $N0;
    border: 1px solid $N30;
    outline: none;
    resize: none;
    &:focus{
      box-shadow: 0 1px 2px 0 rgba(109, 53, 195, 0.15);
      border: solid 1px $P75;
    }
  }
  .textareaName{
    position: relative;
    bottom: 70px;
  }

  .rightShow{
    margin-top: 7px;
    margin-left: 170px;
  }
  .note{
    font-size: 12px;
    color: $N100;
  }
  .submitBtnGroup{
    margin: 0 -32px;
    padding: 16px 0 16px 224px;
    border-top: 1px solid $N10;
    margin-bottom: 0;
    button{
      margin-right: 12px;
    }
  }
}
