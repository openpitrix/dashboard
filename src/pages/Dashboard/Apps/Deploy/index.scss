@import '~scss/vars';

.deployApp {
  min-width: 1128px !important;
  margin: 0 auto;
}

.appBaseInfo {
  margin-bottom: 20px;
  padding-bottom: 20px;
  border-bottom: 1px solid $N10;

  .title {
    margin-bottom: 32px;
  }

  .icon {
    float: left;
    width: 40px;
    height: 40px;
    line-height: 40px;
    margin-right: 16px;

    img {
      max-width: 100%;
      max-height: 100%;
      vertical-align: middle;
    }
  }

  .name {
    @include title-font($line-height: 40px, $font-size: 20px);
    @include textCut;
  }

  .abstraction {
    @include note-font;
  }
}

.typeVersions {
  dl {
    margin-bottom: 20px;
    overflow: hidden;
  }

  dt {
    @include normal-font;
    display: inline-block;
    width: 70px;
    padding: 6px 12px 6px 0;
    text-align: right;
  }

  dd {
    display: inline-block;
    float: right;
    width: calc(100% - 82px);
  }

  .types {
    // margin-bottom: 16px;

    > label {
      display: inline-block;
      padding: 8px 12px;
      border-radius: 2px;
      background-color: $N0;
      border: 1px solid $N10;
      font-size: 14px;
      line-height: 14px;
      color: $N300;
      cursor: pointer;

      &.active {
        color: $P75;
        border-color: $P75;
      }
    }
  }
}

.selectRuntime {
  padding-left: 150px;

  .name {
    @include normal-font($line-height: 32px);
    float: left;
    display: inline-block;
    margin-right: 20px;
  }

  .select {
    float: left;
    width: 326px;
  }

  & > a {
    display: inline-block;
    margin-left: 12px;
    line-height: 32px;
    font-size: 12px;
  }

}

.deployForm {
  padding: 32px;
}

.header {
  margin: 24px 0;
  a {
    line-height: 20px;
    font-weight: 500;
    color: $title-color;
    transition: color 0.3s ease-in-out;
    &:hover {
      color: lighten($title-color, 20%);
    }
  }
}

.createForm {
  display: block;
  margin: 0;
  margin-top: 5px;

  .moduleTitle {
    //margin: 50px 0 25px 0;
    margin-bottom: 28px;
    font-size: 16px;
    color: $N500;
    &:not(:first-of-type) {
      margin-top: 50px;
    }
  }
  .cellModule {
    margin-bottom: 20px;
    min-height: 24px;
    clear: both;
  }
  .name {
    display: inline-block;
    //float: left;
    margin-right: 24px;
    width: 168px;
    font-size: 14px;
    font-weight: 500;
    color: $N500;
    text-align: right;
  }
  .selectName {
    float: left;
    line-height: 32px;
  }
  .radioName {
    float: left;
    margin-top: 5px;
  }
  .showWord {
    display: inline-block;
    max-width: calc(100% - 192px);
    line-height: 32px;
  }

  .radio,
  .checkbox {
    margin-right: 20px;
  }
  .add {
    display: inline-block;
    font-weight: bold;
  }
  .select,
  .input {
    width: 256px;
    margin-right: 12px;
  }
  .textarea {
    box-sizing: border-box;
    padding: 7px 10px;
    width: 400px;
    height: 88px;
    border-radius: 2px;
    background-color: $N0;
    border: 1px solid $N30;
    outline: none;
    resize: none;
    &:focus {
      box-shadow: 0 1px 2px 0 rgba(109, 53, 195, 0.15);
      border: solid 1px $P75;
    }
  }
  .textareaName {
    position: relative;
    bottom: 70px;
  }

  .rightShow {
    margin-top: 7px;
    margin-left: 170px;
  }

  .note {
    font-size: 12px;
    color: $N100;
  }

  .submitBtnGroup {
    margin: 0 -32px;
    padding: 16px 0 16px 224px;
    border-top: 1px solid $N10;
    margin-bottom: 0;
    button {
      margin-right: 12px;
    }
  }

  .createRuntime {
    line-height: 24px;
  }
}

.errMsg {
  max-width: 80%;
  border-radius: 2px;
  background-color: $R10;
  box-shadow: 0 2px 4px 0 rgba(178, 42, 38, 0.1);
  line-height: 1.71;
  color: $R300;
  padding-left: 10px;
}
