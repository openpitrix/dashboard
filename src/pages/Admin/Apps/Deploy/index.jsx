import React, { Component, Fragment } from 'react';
import { observer, inject } from 'mobx-react';
import { toJS } from 'mobx';
import { Link } from 'react-router-dom';
import { Radio, Button, Input, Select, Slider } from 'components/Base';
import Layout, { BackBtn, CreateResource } from 'components/Layout/Admin';
import Cell from './Cell/index.jsx';
import { get } from 'lodash';

import styles from './index.scss';
import classNames from 'classnames';

@inject(({ rootStore }) => ({
  rootStore,
  appStore: rootStore.appStore,
  appDeployStore: rootStore.appDeployStore,
  runtimeStore: rootStore.runtimeStore
}))
@observer
export default class AppDeploy extends Component {
  static async onEnter({ appStore, appDeployStore, runtimeStore }, params) {
    appDeployStore.appId = params.appId;
    await appDeployStore.fetchVersions({ app_id: [params.appId] }, true);
    await appDeployStore.fetchRuntimes();
  }

  render() {
    const { appDeployStore } = this.props;
    const { notifyMsg, hideMsg } = appDeployStore;
    const title = 'Deploy app';
    return (
      <Layout msg={notifyMsg} hideMsg={hideMsg} isLoading={appDeployStore.isLoading}>
        <BackBtn label="clusters" link="/dashboard/clusters" />
        <CreateResource title={title} aside={this.renderAside()}>
          {this.renderForm()}
        </CreateResource>
      </Layout>
    );
  }

  renderAside() {
    return (
      <Fragment>
        <p>
          Application repos are labelled for GUI to show in category list, and have label selector
          to choose which runtime to run when user deploys any application that belongs to the repo.
        </p>
        <p>Runtime env is labelled. A runtime can have multiple labels.</p>
        <p>
          Repo indexer will scan configured repo list periodically and cache the metadata of the
          repos.
        </p>
        <p>Repo manager is responsible for creating/deleting/updating repos.</p>
      </Fragment>
    );
  }

  renderForm() {
    const { appDeployStore, runtimeStore } = this.props;
    const {
      config,
      configBasics,
      configNodes,
      configEnvs,
      changeCell,
      handleSubmit,
      isLoading,
      versions,
      runtimes,
      subnets,
      versionId,
      runtimeId,
      subnetId,
      changeRuntime,
      changeVersion,
      changeSubnet
    } = appDeployStore;

    return (
      <form
        className={styles.createForm}
        method="post"
        onSubmit={handleSubmit.bind(appDeployStore)}
      >
        <div className={styles.moduleTitle}>1. Basic settings</div>
        {configBasics &&
          configBasics.map(
            (basic, index) =>
              basic.key !== 'subnet' && (
                <Cell
                  key={basic.key}
                  className={styles.cellModule}
                  config={basic}
                  type={`basic`}
                  configIndex1={index}
                  configIndex2={-1}
                  changeCell={changeCell.bind(appDeployStore)}
                />
              )
          )}
        <div className={styles.cellModule}>
          <label className={styles.name}>Runtime</label>
          <Radio.Group className={styles.showWord} value={runtimeId} onChange={changeRuntime}>
            {runtimes &&
              runtimes.map(({ runtime_id, name }) => (
                <Radio key={runtime_id} value={runtime_id}>
                  {name}
                </Radio>
              ))}
          </Radio.Group>
        </div>
        <div className={styles.cellModule}>
          <label className={styles.name}>Version</label>
          <Select className={styles.select} value={versionId} onChange={changeVersion}>
            {versions.map(({ version_id, name }) => (
              <Select.Option key={version_id} value={version_id}>
                {name}
              </Select.Option>
            ))}
          </Select>
        </div>

        {configNodes &&
          configNodes.map((node, index1) => (
            <Fragment key={node.key}>
              <div className={styles.moduleTitle}>
                {index1 + 2}. Node Settings({node.key})
              </div>
              {node.properties.map((conf, index2) => (
                <Cell
                  key={conf.key}
                  className={styles.cellModule}
                  config={conf}
                  type={`node`}
                  configIndex1={index1}
                  configIndex2={index2}
                  changeCell={changeCell.bind(appDeployStore)}
                />
              ))}
            </Fragment>
          ))}

        <div className={styles.moduleTitle}>{configNodes.length + 2}. Vxnet Settings</div>
        <div className={styles.cellModule}>
          <label className={styles.name}>VxNet to Join</label>
          <Select className={styles.select} value={subnetId} onChange={changeSubnet}>
            {subnets.map(({ subnet_id, name }) => (
              <Select.Option key={subnet_id} value={subnet_id}>
                {subnet_id}
              </Select.Option>
            ))}
          </Select>
        </div>

        {configEnvs &&
          configEnvs.map((env, index1) => (
            <Fragment key={env.key}>
              <div className={styles.moduleTitle}>
                {index1 + configNodes.length + 3}.Environment Settings({env.key})
              </div>
              {env.properties.map((conf, index2) => (
                <Cell
                  key={conf.key}
                  className={styles.cellModule}
                  config={conf}
                  type={`env`}
                  configIndex={index1}
                  configIndex2={index2}
                  changeCell={changeCell.bind(appDeployStore)}
                />
              ))}
            </Fragment>
          ))}

        <div className={styles.submitBtnGroup}>
          <Button type={`primary`} className={`primary`} htmlType="submit" disabled={isLoading}>
            Confirm
          </Button>
          <Button
            onClick={() => {
              history.back();
            }}
          >
            Cancel
          </Button>
        </div>
      </form>
    );
  }
}
