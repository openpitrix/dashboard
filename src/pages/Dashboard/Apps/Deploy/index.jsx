import React, { Component } from 'react';
import { Link } from 'react-router-dom';
import { observer, inject } from 'mobx-react';
import classnames from 'classnames';
import { translate } from 'react-i18next';
import _ from 'lodash';

import { Button, Image } from 'components/Base';
import Layout, { BackBtn, CreateResource, BreadCrumb } from 'components/Layout';

import { Group as DeployGroup } from 'components/Deploy';
import VMParser from 'lib/config-parser/vm';
import { getFormData } from 'utils';

import styles from './index.scss';

const validKeyPrefix = ['cluster', 'node', 'env'];
const keysShouldBeNumber = [
  'cpu',
  'memory',
  'storage_size',
  'volume_size',
  'instance_class',
  'count'
];

@translate()
@inject(({ rootStore }) => ({
  rootStore,
  appStore: rootStore.appStore,
  repoStore: rootStore.repoStore,
  appDeployStore: rootStore.appDeployStore,
  runtimeStore: rootStore.runtimeStore,
  user: rootStore.user
}))
@observer
export default class AppDeploy extends Component {
  constructor(props) {
    super(props);
    this.vmParser = new VMParser();
  }

  async componentDidMount() {
    const {
      appStore, repoStore, appDeployStore, user, match
    } = this.props;
    const { appId } = match.params;

    await appStore.fetch(appId);
    if (appStore.appDetail.repo_id) {
      await repoStore.fetchRepoDetail(appStore.appDetail.repo_id);
    }

    const repoProviders = _.get(repoStore.repoDetail, 'providers', []);
    const isK8s = repoProviders.includes('kubernetes');
    appDeployStore.isK8s = isK8s;

    // fetch runtimes
    await appDeployStore.fetchRuntimes({
      status: 'active',
      label: repoStore.getStringFor('selectors'),
      provider: repoProviders,
      owner: user.user_id
    });

    if (!isK8s) {
      appDeployStore.runtimeId = _.get(
        appDeployStore.runtimes.slice(),
        '[0].runtime_id'
      );
      await appDeployStore.fetchSubnetsByRuntime(appDeployStore.runtimeId);
    }

    // fetch versions
    await appDeployStore.fetchVersions({ app_id: [appId] });
    appDeployStore.versionId = _.get(
      appDeployStore.versions.slice(),
      '[0].version_id'
    );
    await appDeployStore.fetchFilesByVersion(appDeployStore.versionId, isK8s);

    if (!isK8s && !_.isEmpty(appDeployStore.configJson)) {
      this.vmParser.setConfig(appDeployStore.configJson);
      this.forceUpdate();
    }
  }

  componentWillUnmount() {
    this.props.appDeployStore.reset();
  }

  handleSubmit = async e => {
    e.preventDefault();

    const {
      appDeployStore, user, history, match, t
    } = this.props;
    const { isK8s, create } = appDeployStore;

    const formData = this.getFormDataByKey();
    const name = _.get(formData, 'cluster.name', '');
    if (!name) {
      return appDeployStore.error(t('Name should not be empty'));
    }

    let conf;
    if (isK8s) {
      const yamlStr = _.get(formData, 'values.yaml', '');
      if (!yamlStr) {
        return appDeployStore.error(t('Invalid yaml'));
      }
      conf = [`Name: ${name}`, yamlStr].join('\n').replace(/#.*/g, '');
    } else {
      conf = JSON.stringify({
        cluster: _.extend(
          {},
          this.getFormDataByKey('cluster'),
          this.getNodesConf()
        ),
        env: this.getFormDataByKey('env')
      });
    }

    const versionId = formData['cluster.version'] || '';
    const runtimeId = formData['cluster.runtime'] || '';

    if (!versionId) {
      return appDeployStore.info(t('Version should not be empty'));
    }
    if (!runtimeId) {
      return appDeployStore.info(t('Runtime should not be empty'));
    }

    const res = await create({
      app_id: match.params.appId,
      version_id: versionId,
      runtime_id: runtimeId,
      conf: conf.replace(/>>>>>>/g, '.')
    });

    if (res && _.get(res, 'cluster_id')) {
      appDeployStore.success(t('Deploy app successfully'));
      const path = user.isNormal ? '/purchased' : '/dashboard/clusters';
      setTimeout(() => history.push(path), 1000);
    }
  };

  handleCancel = () => {
    this.props.history.go(-1);
  };

  getFormDataByKey = (keyPrefix = '') => {
    const formData = getFormData(this.deployForm);

    if (!keyPrefix) {
      return formData;
    }

    if (validKeyPrefix.indexOf(keyPrefix) < 0) {
      throw Error(`invalid keyPrefix: ${keyPrefix}`);
    }

    if (!keyPrefix.endsWith('.')) {
      keyPrefix += '.';
    }

    const dataByPrefix = _.pickBy(
      formData,
      (val, key) => key.indexOf(keyPrefix) === 0
    );

    return _.mapKeys(dataByPrefix, (val, key) => key.substring(keyPrefix.length));
  };

  getNodesConf = () => {
    const nodesConf = this.getFormDataByKey('node');

    return _.transform(
      nodesConf,
      (res, val, key) => {
        const [role, meter] = key.split('.');
        res[role] = res[role] || {};
        res[role][meter] = keysShouldBeNumber.indexOf(meter) > -1 ? parseInt(val) : val;
      },
      {}
    );
  };

  renderAside() {
    const { appStore, t } = this.props;
    const { appDetail } = appStore;

    return (
      <div className={styles.aside}>
        <div className={styles.appIntro}>
          <Image
            src={appDetail.icon}
            className={styles.icon}
            iconLetter={appDetail.name}
          />
          <span className={styles.name}>{appDetail.name}</span>
        </div>
        <p>{t('Deploy App explain')}</p>
      </div>
    );
  }

  renderForEmptyItem = itemKey => {
    const { t } = this.props;

    if (!this.isDeployReady()) {
      return null;
    }

    if (itemKey === 'cluster.runtime') {
      return (
        <span>
          {t('DEPLOY_NO_RUNTIME_NOTE')}
          <Link to={'/dashboard/runtime/create'}>{t('create')}</Link>
        </span>
      );
    }
  };

  renderBody() {
    if (!this.isDeployReady()) {
      return null;
    }

    const { appDeployStore, t } = this.props;
    const {
      isK8s,
      yamlStr,
      normalizeRuntime,
      normalizeVersions,
      normalizeSubnets,
      onChangeFormField,
      runtimeId,
      versionId
    } = appDeployStore;

    let groups = [];

    if (!isK8s) {
      if (!this.vmParser.isReady()) {
        return null;
      }

      this.vmParser.setSubnets(normalizeSubnets());
    }

    // always set runtimes and versions
    this.vmParser.setRuntimes(normalizeRuntime(), { default: runtimeId });
    this.vmParser.setVersions(normalizeVersions(), { default: versionId });

    if (isK8s) {
      groups = [].concat(
        {
          title: t('Basic settings'),
          items: this.vmParser.getDefaultBasicSetting({
            key: 'name',
            description: t('HELM_APP_NAME_TIP')
          })
        },
        {
          title: t('Deploy settings'),
          items: [
            {
              type: 'string',
              renderType: 'yaml',
              keyName: 'values.yaml',
              yaml: yamlStr
            }
          ]
        }
      );
    } else {
      const basicSetting = this.vmParser.getBasicSetting();
      const nodeSetting = this.vmParser.getNodeSetting();
      const envSetting = this.vmParser.getEnvSetting();
      const vxnetSetting = this.vmParser.getVxnetSetting();

      groups = [].concat(
        { title: t('Basic settings'), items: basicSetting },
        nodeSetting,
        { title: t('Vxnet settings'), items: vxnetSetting },
        envSetting
      );
    }

    return groups.map((group, idx) => (
      <DeployGroup
        detail={group}
        seq={idx}
        key={idx}
        onChange={onChangeFormField}
        onEmpty={this.renderForEmptyItem}
      />
    ));
  }

  isDeployReady() {
    const { configJson, yamlStr } = this.props.appDeployStore;
    return !_.isEmpty(configJson) || `${yamlStr}`;
  }

  renderFooter() {
    const { t } = this.props;
    const {
      isLoading,
      isK8s,
      runtimes,
      versions,
      subnets
    } = this.props.appDeployStore;

    if (!this.isDeployReady()) {
      return null;
    }

    let canSubmit = runtimes.length && versions.length;

    if (!isK8s) {
      canSubmit = canSubmit && subnets.length;
    }

    return (
      <div className={styles.actionBtnGroup}>
        <Button
          type="primary"
          onClick={this.handleSubmit}
          className={styles.btn}
          disabled={isLoading || !canSubmit}
        >
          {t('Confirm')}
        </Button>
        <Button onClick={this.handleCancel} className={styles.btn}>
          {t('Cancel')}
        </Button>
      </div>
    );
  }

  render() {
    const {
      appDeployStore, appStore, user, t
    } = this.props;
    const { appDetail } = appStore;
    const { isLoading, errMsg } = appDeployStore;

    const title = `${t('Deploy')} ${appDetail.name}`;
    const { isNormal, isDev } = user;
    const linkPath = isDev ? 'My Apps>Test>Deploy' : 'Store>All Apps>Deploy';

    return (
      <Layout
        className={classnames({ [styles.deployApp]: !isNormal })}
        title="Store"
        hasSearch
        isLoading={isLoading}
        backBtn={
          isNormal && (
            <BackBtn
              label={appDetail.name}
              link={`/store/${appDetail.app_id}`}
            />
          )
        }
      >
        {!isNormal && <BreadCrumb linkPath={linkPath} />}

        <CreateResource
          title={title}
          aside={this.renderAside()}
          asideTitle=""
          footer={this.renderFooter()}
        >
          {errMsg && <p className={styles.errMsg}>{t(errMsg)}</p>}
          <form
            className={styles.createForm}
            method="post"
            onSubmit={this.handleSubmit}
            ref={node => {
              this.deployForm = node;
            }}
          >
            {this.renderBody()}
          </form>
        </CreateResource>
      </Layout>
    );
  }
}
