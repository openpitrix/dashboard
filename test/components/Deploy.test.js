import Deploy from 'components/Deploy';

describe('Component Deploy ', () => {
  it('basic render', () => {
    const wrapper = render(<Deploy />);

    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
