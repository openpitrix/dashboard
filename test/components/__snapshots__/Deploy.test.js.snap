// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component Deploy  basic render 1`] = `
<div
  class="row"
>
  <label
    class="label"
  />
  <div
    class="item"
  />
</div>
`;
